File: lib/shared/logic/bid_validator.dart
import '../models/bid.dart';

enum BidValidationError {
  alreadyPassed,
  bidNotHigher,
  mustBid;

  @override
  String toString() => switch (this) {
        BidValidationError.alreadyPassed => 'already-passed',
        BidValidationError.bidNotHigher => 'bid-not-higher',
        BidValidationError.mustBid => 'must-bid',
      };
}

class BidValidationResult {
  final bool isValid;
  final BidValidationError? error;
  const BidValidationResult.valid() : isValid = true, error = null;
  const BidValidationResult.invalid(this.error) : isValid = false;
}

class BiddingOutcome {
  final bool isComplete;
  final int? winnerIndex;
  final BidAmount? winningBid;

  const BiddingOutcome._({required this.isComplete, this.winnerIndex, this.winningBid});

  factory BiddingOutcome.won({required int winnerIndex, required BidAmount bid}) =>
      BiddingOutcome._(isComplete: true, winnerIndex: winnerIndex, winningBid: bid);

  factory BiddingOutcome.ongoing() => const BiddingOutcome._(isComplete: false);

  @override
  bool operator ==(Object other) {
    if (other is! BiddingOutcome) return false;
    return isComplete == other.isComplete && winnerIndex == other.winnerIndex && winningBid == other.winningBid;
  }

  @override
  int get hashCode => Object.hash(isComplete, winnerIndex, winningBid);
}

class BidValidator {
  static bool _alreadyPassed(List<int> passedPlayers, int playerIndex) =>
      passedPlayers.contains(playerIndex);

  static BidValidationResult validateBid({
    required BidAmount bidAmount,
    required BidAmount? currentHighest,
    required List<int> passedPlayers,
    required int playerIndex,
  }) {
    if (_alreadyPassed(passedPlayers, playerIndex)) {
      return const BidValidationResult.invalid(BidValidationError.alreadyPassed);
    }
    if (currentHighest != null && bidAmount.value <= currentHighest.value) {
      return const BidValidationResult.invalid(BidValidationError.bidNotHigher);
    }
    return const BidValidationResult.valid();
  }

  static BidValidationResult validatePass({
    required List<int> passedPlayers,
    required int playerIndex,
    int playerCount = 4,
    BidAmount? currentHighest,
  }) {
    if (_alreadyPassed(passedPlayers, playerIndex)) {
      return const BidValidationResult.invalid(BidValidationError.alreadyPassed);
    }
    if (isLastBidder(passedPlayers: passedPlayers, playerIndex: playerIndex, playerCount: playerCount) && currentHighest == null) {
      return const BidValidationResult.invalid(BidValidationError.mustBid);
    }
    return const BidValidationResult.valid();
  }

  static bool isLastBidder({
    required List<int> passedPlayers,
    required int playerIndex,
    int playerCount = 4,
  }) {
    if (_alreadyPassed(passedPlayers, playerIndex)) return false;
    final activePlayers = playerCount - passedPlayers.length;
    return activePlayers == 1;
  }

  static BiddingOutcome checkBiddingComplete({
    required List<int> passedPlayers,
    required BidAmount? currentHighest,
    required int? highestBidderIndex,
  }) {
    if (passedPlayers.length >= 3 && currentHighest != null && highestBidderIndex != null) {
      return BiddingOutcome.won(winnerIndex: highestBidderIndex, bid: currentHighest);
    }
    return BiddingOutcome.ongoing();
  }
}

File: workers/src/game/bid-validator.ts
export interface BidValidationResult {
  valid: boolean;
  error?: string;
}

export interface BiddingCompleteResult {
  complete: boolean;
  winner?: string;
  bid?: number;
}

export function validateBid(
  bidAmount: number,
  currentHighest: number | null,
  passedPlayers: string[],
  playerId: string
): BidValidationResult {
  if (passedPlayers.includes(playerId)) {
    return { valid: false, error: 'already-passed' };
  }
  if (currentHighest !== null && bidAmount <= currentHighest) {
    return { valid: false, error: 'bid-not-higher' };
  }
  return { valid: true };
}

export function isLastBidder(
  passedPlayers: string[],
  playerId: string,
  playerCount: number = 4
): boolean {
  if (passedPlayers.includes(playerId)) return false;
  const activePlayers = playerCount - passedPlayers.length;
  return activePlayers === 1;
}

export function validatePass(
  passedPlayers: string[],
  playerId: string,
  playerCount: number = 4,
  currentHighest: number | null = null
): BidValidationResult {
  if (passedPlayers.includes(playerId)) {
    return { valid: false, error: 'already-passed' };
  }
  if (isLastBidder(passedPlayers, playerId, playerCount) && currentHighest === null) {
    return { valid: false, error: 'must-bid' };
  }
  return { valid: true };
}

export function checkBiddingComplete(
  passedPlayers: string[],
  currentHighest: number | null,
  highestBidder: string | null
): BiddingCompleteResult {
  if (passedPlayers.length >= 3 && currentHighest !== null && highestBidder !== null) {
    return { complete: true, winner: highestBidder, bid: currentHighest };
  }
  return { complete: false };
}
